Source/Core/Core/PowerPC/Jit64/Jit_LoadStore.cpp

```diff
@@ -234,21 +234,52 @@ void Jit64::dcbx(UGeckoInstruction inst)
   JITDISABLE(bJITLoadStoreOff);
 
   X64Reg addr = RSCRATCH;
+  X64Reg value = RSCRATCH2;
   RCOpArg Ra = inst.RA ? gpr.Use(inst.RA, RCMode::Read) : RCOpArg::Imm32(0);
   RCOpArg Rb = gpr.Use(inst.RB, RCMode::Read);
-  RegCache::Realize(Ra, Rb);
+  RCX64Reg tmp = gpr.Scratch();
+  RegCache::Realize(Ra, Rb, tmp);
 
-  MOV_sum(32, addr, Ra, Rb);
-  AND(32, R(addr), Imm8(~31));
+  // Translate effective address to physical address.
+  MOV_sum(32, value, Ra, Rb);
+  FixupBranch bat_lookup_failed;
+  if (MSR.IR)
+  {
+    MOV(32, R(addr), R(value));
+    bat_lookup_failed = BATAddressLookup(value, tmp);
+    AND(32, R(addr), Imm32(0x0001ffff));
+    AND(32, R(value), Imm32(0xfffe0000));
+    OR(32, R(value), R(addr));
+  }
+  MOV(32, R(addr), R(value));
+
+  // Check whether a JIT cache line needs to be invalidated.
+  SHR(32, R(value), Imm8(5 + 5));  // >> 5 for cache line size, >> 5 for width of bitset
+  MOV(64, R(tmp), ImmPtr(GetBlockCache()->GetBlockBitSet()));
+  MOV(32, R(value), MComplex(tmp, value, SCALE_4, 0));
+  SHR(32, R(addr), Imm8(5));
+  BT(32, R(value), R(addr));
```

Comment:
The other thing I'm wondering here if this needs an extra `AND(32, R(addr), 0x1F)`? I suspect not as references online claim that `bt` does an implicit mod32 on the 2nd parameter, which seems logical, but curiously MSVC does emit this `and` in the bitset code, even in Release mode. Oh well, I think this is fine as-is.

Reply:
Yeah, x64 clears the upper bits of the register when you do a 32bit operation. 

This is fine.